File: api-rest/hotel_cervera_api/src/main/java/com/hotel/cervera/hotel_cervera_api/model/Gasto.java
package com.hotel.cervera.hotel_cervera_api.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "gastos")
@Getter @Setter @NoArgsConstructor @AllArgsConstructor @Builder
public class Gasto {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "fecha_gasto", nullable = false)
    private LocalDate fechaGasto;

    @Column(name = "descripcion", nullable = false, length = 200)
    private String descripcion;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "categoria_id", nullable = false)
    private CategoriaGasto categoria;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "tipo_gasto_id", nullable = false)
    private TipoGasto tipoGasto;

    @Column(name = "monto", nullable = false, precision = 10, scale = 2)
    private BigDecimal monto;

    @Column(name = "observaciones", columnDefinition = "TEXT")
    private String observaciones;

    @Builder.Default
    @Column(name = "estado", nullable = false, length = 20)
    private String estado = "ACTIVO"; // ACTIVO, ANULADO

    // Auditoría de Creación
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "creado_por", nullable = false)
    private Usuario creadoPor;

    @CreationTimestamp
    @Column(name = "fecha_creacion", updatable = false)
    private OffsetDateTime fechaCreacion;

    // Auditoría de Actualización
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "actualizado_por")
    private Usuario actualizadoPor;

    @UpdateTimestamp
    @Column(name = "fecha_actualizacion")
    private OffsetDateTime fechaActualizacion;

    // Auditoría de Anulación
    @Column(name = "fecha_anulacion")
    private OffsetDateTime fechaAnulacion;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "anulado_por")
    private Usuario anuladoPor;

    @Column(name = "motivo_anulacion", columnDefinition = "TEXT")
    private String motivoAnulacion;
}

File: web-app/src/app/core/models/gasto.ts
export interface CategoriaGasto {
  id: number;
  nombre: string;
  descripcion?: string;
  activo: boolean;
  orden: number;
}

export interface TipoGasto {
  id: number;
  nombre: string;
}

export interface GastoResponse {
  id: string;
  fechaGasto: string;
  descripcion: string;
  
  categoriaId: number;
  categoriaNombre: string;
  
  tipoGastoId: number;
  tipoGastoNombre: string;
  
  monto: number;
  observaciones?: string;
  estado: string; // 'ACTIVO' | 'ANULADO'
  
  creadoPor: string;
  creadoPorNombre: string;
  fechaCreacion: string;
  
  actualizadoPor?: string;
  actualizadoPorNombre?: string;
  fechaActualizacion: string;
  
  fechaAnulacion?: string;
  anuladoPor?: string;
  anuladoPorNombre?: string;
  motivoAnulacion?: string;
}

export interface GastoRequest {
  fechaGasto: string;
  descripcion: string;
  categoriaId: number;
  tipoGastoId: number;
  monto: number;
  observaciones?: string;
  creadoPor: string;
}
